### ISSUE TITLE
feat(moe): leaf-value refit-on-r-update — root-cause fix for additive-only EM (v0.7)

### ISSUE THREAD

**Issue by kyo219** [kyo219/LightGBM-MoE#37]
## Background

The MoE training loop in `MixtureGBDT::TrainOneIter` (`src/boosting/mixture_gbdt.cpp:2880`) is **Generalized EM (GEM) with a single functional-gradient step per outer iteration** — each round only *appends* one tree per expert, with leaf values determined by the responsibilities `r_ik` at the moment of creation. Historical trees and historical leaf values are frozen.

This produces the limitation called out in the README's "EM updates are additive-only" section: large regime re-assignments cannot happen mid-training, because

- `f_k(x)` only changes by `η · single_tree(x)` per round → bounded per-iter shift
- new `r` ⇒ small change in `loss(y_i, f_k(x_i))` ⇒ small change in `r` next round
- the basin set by `r_init` is sticky; the gate's accumulated trees lock in routing toward whichever expert `r` concentrated on early

The root cause, when traced carefully, **is not "we only add trees" — it is "leaf values are write-once, fixed at the `r` of the round they were built in"**. The tree partitions form a useful data-dependent feature representation `Φ_k(x)`; the leaf values are linear coefficients on top of `Φ_k`. In classical EM, the linear coefficients have a closed-form solution under the current `r` and are fully rewritten each iteration. In our current implementation they are never rewritten.

A PR-#36 sin-perturbation symmetry breaker, the PR-#26 Huber-clipped diversity reward, and the PR-#24 variance estimator are all *symptomatic* fixes for downstream consequences of frozen leaf values. Refitting the leaves directly on each E-step would resolve the same problem at the source.

## Goal

After the E-step recomputes `r`, refit the leaf values of all (or recent, or trigger-selected) expert trees in closed form against the new `r`-weighted gradient/Hessian, **before** appending the next tree. This restores the property that `f_k(x)` is the r-conditional optimum on its current partition structure — i.e. the actual M-step solution rather than a single descent step.

The expected effect:

- `f_k(x)` can take large jumps between rounds when `r` shifts
- `r` can take correspondingly large jumps in the next E-step
- ELBO becomes much closer to monotone non-decreasing (currently we log warnings on >5% drops; with refit those should disappear except in pathological configs)
- "EM flipped two components" behavior — observable in `model.get_responsibilities()` snapshots — becomes possible mid-training

## Proposed approach

1. **New config flags** in `include/LightGBM/config.h`:
   ```
   bool   mixture_refit_leaves      = false;          // opt-in for v0.7
   string mixture_refit_trigger     = "elbo";         // "always" | "elbo" | "every_n"
   int    mixture_refit_every_n     = 10;             // for "every_n" trigger
   int    mixture_refit_window      = -1;             // refit only most-recent N trees; -1 = all
   ```

2. **Leaf-assignment cache** on `MixtureGBDT`: `vector<vector<int>> leaf_assignment_[K][tree_id]` where `[k][t][i]` is the leaf index sample `i` falls into in expert `k`'s tree `t`. Built once when a tree is added; invalidated on tree-shape changes (currently never — trees are immutable once added).

3. **Refit dispatch** in `TrainOneIter` between E-step and `MStepExperts`:
   ```cpp
   if (config_->mixture_refit_leaves && ShouldRefit()) {
       RefitExpertLeaves();   // new method
       RefitGateLeaves();     // also refit gate trees — same root cause
   }
   ```
   `ShouldRefit()` evaluates the trigger:
   - `always`: every iter past warmup
   - `elbo`: when `prev_marginal_log_lik_ - current_ll > 0.02 * |prev|` (re-use existing ELBO machinery in `mixture_gbdt.cpp:1808`)
   - `every_n`: every N rounds

4. **`RefitExpertLeaves()`** — for each expert `k`, for each tree `t` (within `mixture_refit_window` of newest):
   - For each leaf `l`: aggregate `Σ_{i ∈ leaf_l} r_ik · g_i` and `Σ_{i ∈ leaf_l} r_ik · h_i` using the cached leaf assignment
   - Set new leaf value `v_{k,t,l} ← -η · sum_grad_l / (sum_hess_l + λ)`
   - Use existing LightGBM `Tree::SetLeafOutput` (already public) — no need for a new core API
   - The `experts_[k]->RefitTree(...)` core API can be used as a fallback if SetLeafOutput direct manipulation is unsafe

5. **`RefitGateLeaves()`** — same logic against soft-CE target `r`. Per-class trees of the multiclass GBDT each get their leaf values refit using the gate's own gradient form `(p − r)/T` and Hessian `K/(K−1)·p(1−p)/T²` (matches `MStepGate`).

6. **Diagnostic instrumentation**:
   - Log `||r_t − r_{t-1}||_F` per round (already partially covered by ELBO; expose explicitly)
   - Log refit invocations and which trigger fired
   - When `mixture_refit_trigger="elbo"`, count how often it fires across a study — should be rare (maybe 5-20 times in 500 iters) to validate cost claims

## Files likely to touch

- `include/LightGBM/config.h` — new mixture_refit_* flags
- `src/boosting/mixture_gbdt.h` — declare RefitExpertLeaves / RefitGateLeaves / ShouldRefit, leaf-assignment cache members
- `src/boosting/mixture_gbdt.cpp` — implement above; hook into `TrainOneIter` between E-step and `MStepExperts`
- `python-package/lightgbm_moe/basic.py` — no API surface change; flags pass through `params`
- `tests/python_package_test/test_engine.py` — new test: `test_moe_refit_leaves_enables_large_r_jumps`
- `examples/em_refit_demo.py` — new: synthetic dataset where `r_init` is intentionally bad; show that with refit on, `r` recovers; with refit off, it's stuck (tracking via `get_responsibilities` callback)
- `docs/moe/architecture.md` — new §2.7 "Refit-leaves on E-step (v0.7)"
- `README.md` — update Limitations section: "as of v0.6.0" → "before v0.7's `mixture_refit_leaves` flag"

## Acceptance criteria

- **Mechanism check**: `examples/em_refit_demo.py` shows `||r_t − r_init||` reaching ≥0.3 within 50 iters when refit is on, vs <0.05 when off, on the same synthetic regime-flip setup.
- **ELBO monotonicity**: on the 500-trial study's 6 datasets, the count of >5% ELBO drops decreases by ≥80% with `mixture_refit_leaves=true, mixture_refit_trigger=always`.
- **No regression on default path**: with `mixture_refit_leaves=false` (default), bench numbers on the v0.6.0 6-row study are bit-identical (the flag is purely additive).
- **Held-out RMSE**: with `mixture_refit_trigger=elbo` (the recommended setting), held-out RMSE on the regime-observable rows (`synthetic`, `vix`, `hmm`) does not regress; ideally improves on `fred_gdp` where the gate currently loses to the K-way ensemble (the small-data basin escape is exactly what refit should help with).
- **Cost**: `elbo` trigger fires ≤30 times per 500-iter run on average across the 6 datasets. `always` trigger is ≤3× wall-clock vs `false` at 500 iters / 4 experts.

## Pre-conditions / scope

- 2-3 week single PR; should be split into commits: (a) leaf-assignment cache + plumbing, (b) `RefitExpertLeaves`, (c) `RefitGateLeaves`, (d) trigger logic + ELBO integration, (e) tests + bench + docs.
- Bench against v0.6.0 release tag to confirm no regression on the default-off path.
- Optuna search over `mixture_refit_*` on a single dataset (`fred_gdp`, since that's where the additive-only limitation hurts most) before promoting to the headline study.
- v0.7 release plan should not bundle other major behavior changes — refit is structural enough to be its own release.

## References

- Limitations section in [`README.md`](https://github.com/kyo219/LightGBM-MoE/blob/master/README.md) — "EM updates are additive-only (model-specific)"
- [`docs/moe/architecture.md`](https://github.com/kyo219/LightGBM-MoE/blob/master/docs/moe/architecture.md) §2.3 (M-step on experts) and §2.4 (M-step on gate)
- Existing in-place expert mutation precedent: `SpawnExpertsFromSeed` (`src/boosting/mixture_gbdt.cpp:1391`) — proves we already manipulate `experts_[k]` post-init via `SetLeafValue`
- LightGBM core: `Boosting::RefitTree` API in `include/LightGBM/boosting.h`
- ELBO monotonicity infrastructure already present: `ComputeMarginalLogLikelihood` (`src/boosting/mixture_gbdt.cpp:1808`) and the >5% drop warning in `TrainOneIter`
- Symptomatic fixes that refit-leaves subsumes / makes redundant: PR #36 (uniform-r symmetry breaker), PR #26 (diversity reg Huber clip), parts of PR #24 (variance estimator anti-collapse)
- Design conversation: session notes 2026-05-02, "fundamental fix to additive-only EM" thread
